Android download rewrite

nichwall opened the pull request:
<!--
For Work In Progress Pull Requests, please use the Draft PR feature,
see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

If you do not follow this template, the PR may be closed without review.

Please ensure all checks pass.
If you are a new contributor, the workflows will need to be manually approved before they run.
-->

## Brief summary

This PR refactors the Android download logic to use a single code path for both Internal App Storage and Shared Storage. This PR also allows for resuming an incomplete download.

**This PR does not add support for downloading with the app in the background (see Further Work below).**

## Which issue is fixed?

Fixes https://github.com/advplyr/audiobookshelf-app/issues/862
Fixes https://github.com/advplyr/audiobookshelf-app/issues/34

This PR is mostly just refactoring without fixing any specific issues. I asked for feedback in the following issues to see if the refactor fixed the issue, but still unsure if those issues are fixed so we should leave them open.

Downloading:
- https://github.com/advplyr/audiobookshelf-app/issues/1479
- https://github.com/advplyr/audiobookshelf-app/issues/1273
- https://github.com/advplyr/audiobookshelf-app/issues/1253
- https://github.com/advplyr/audiobookshelf-app/issues/1167
- https://github.com/advplyr/audiobookshelf-app/issues/215

Associating downloaded media with server connection:
- https://github.com/advplyr/audiobookshelf-app/issues/1386
- https://github.com/advplyr/audiobookshelf-app/issues/1362
- https://github.com/advplyr/audiobookshelf-app/issues/594

## Pull Request Type
Android backend

## In-depth Description
This PR is a continuation of https://github.com/advplyr/audiobookshelf-app/pull/1469. Previously, there were two nearly identical code paths for Internal App Storage (using `File`) and Shared Storage (using `DocumentFile`). We also performed an expensive scan (long running) in Shared Storage once all of the files completed downloading.

As part of this PR, we now:
- Always download to Internal App Storage, then move to Shared Storage if desired.
  - This allows us to reuse the same download logic for both internal and shared storage (using `InternalDownloadManager`). This also simplifies continuing incomplete downloads, discussed below.
- `LocalFile` class updates:
  - Remove `simplePath` because it is one of several slow properties (SAF) and is unused anywhere in the codebase.
  - Added `extension` to improve code reuse.
- Build the `LocalLibraryItem` object from the `DownloadItem` object instead of scanning folder in Shared Storage.
  - This still takes about 50-100ms per file in my testing, but is much faster than scanning the folder and *then* still building the object through SAF.
- Say "Moving and processing files" instead of just "Processing"
  - More descriptive about what is happening. When downloading a single large file to Shared Storage, moving the file takes a long time and processing is very quick. When downloading multiple smaller files, moving is very quick (every file is moved as soon as it finishes downloading), but scanning can take a few seconds. Downloading to Internal App Storage does not require moving files and scans many files almost instantly.
- Do not delete temporary or destination files.
  - This allows continuing a long-running download (large single-file M4B), or multi-file downloads.
  - Every file is still added to the download queue. If the file already exists in Internal Storage, continue downloading from the end of the file. If the file is already complete, then the file completes immediately. If the file already exists in Shared Storage, it was already successfully downloaded and moved and is marked as moved.
  - This also allows users to "redownload" a book after resetting the app or manually creating the file structure without actually downloading the entire book again.
  - This drew inspiration from https://github.com/advplyr/audiobookshelf-app/pull/1230 and https://github.com/advplyr/audiobookshelf-app/pull/1231.
- Check free space in Internal App Storage before downloading any file
  - The "minimum free space" required is the size of the downloaded file multiplied by the `maxSimultaneousDownloads`
  - If there is not enough free space, the AbsLogger shows an error and the file is not downloaded. The download queue can continue with other files, so larger files can still be downloaded in series instead of parallel.

### Further work
Building the `LocalLibraryItem` as each individual file is moved to Shared Storage would reduce the processing time to nearly constant (time to move file + 100 ms), but complicates the logic a lot more with incomplete downloads and the queue system.

I would like to move the download manager to a Foreground service to help with downloads being stopped in the background when a user switches to another app. Further testing would be needed before adding a Foreground service. https://github.com/advplyr/audiobookshelf-app/issues/956

## How have you tested this?

Tested with Android 15 on a Pixel 6a with both books and podcasts.
*This should have additional testing done with other devices, especially older versions of Android.*

## Screenshots
N/A

nichwall committed:
Delet unused file

nichwall committed:
Auto formatting `AbsDownloader.kt`

nichwall committed:
Refactor `AbsDownloader`

nichwall committed:
Remove: unused json file from FolderScanner

nichwall committed:
Simplify local library item creation in folder scanner

nichwall committed:
Revert "Refactor `AbsDownloader`"

This reverts commit 4a09238a8e13409e4052b4bfb0c42843b46d333d.

nichwall committed:
Simplify `AbsDownloader` download item creation

nichwall committed:
Combine cover download logic

nichwall committed:
Fix: cover printout for book

nichwall committed:
Initial internal download continue incomplete download

nichwall committed:
Always download to internal app storage

nichwall committed:
Moving folder to shared storage

nichwall committed:
Build direct path to shared storage folder

nichwall committed:
Autoformat `DeviceClasses.kt`

nichwall committed:
Remove: unused simplePath from `LocalFile` object

nichwall committed:
Speed up duplicate docFile calls by using `localFile` reference

nichwall committed:
Merge branch 'master' into android_download_rewrite

nichwall committed:
Add: function to create LocalFile from download item part

nichwall committed:
Combine internal and shared scanner

nichwall committed:
Remove extra printouts

nichwall committed:
Update download complete string

nichwall committed:
Merge branch 'master' into android_download_rewrite

nichwall committed:
Merge branch 'master' into android_download_rewrite

nichwall committed:
Prevent duplicate files from being created and downloaded

timwhite commented:
I find both the 0.9.81-beta and the APK attached to this PR have the same issue, that if you switch away from the app and back, the downloads appear to have frozen and don't resume.

nichwall commented:
Yeah, this PR does not implement downloading in the background. This PR *should* allow you to press the download button again and have the download continue from where it left off, but it sounds like that is not working either. Thanks for the feedback y'all. :)

troy117 commented:
After doing more testing it does continue the download from where it left off, after the app is force closed and then after the download is restarted.  I made one change to your code for the version I am running locally on my phone which is to catch the random end of stream exception that happens and continue the download and it works great.

timwhite commented:
It does work for pressing download again and it resumes part way through a downloaded file, or completed files download basically instantly. So it works for that issue. That in itself makes a huge difference for poor connections, as long as I keep it active.

https://github.com/advplyr/audiobookshelf-app/issues/956 looks to be the best place to track the background/sleep download issue.

nichwall commented:
Ah, cool. Thanks for clarifying that pressing the download button again is working to continue the download (that was the intention).

I'm hoping to address downloading in the background using Work Managers or similar, but I am trying to keep this PR to refactoring with minimal functionality changes to make the following PRs simpler.

nichwall commented:
> Let's say I am downloading 4 audiobooks for holidays. My phone has low internal storage, so I have an SD card for media or other large files (3 GB free internal, 10 GB free card).
> If each book has 1 GB for example, I turn the download on all of them it might fail because the app would fill up the internal storage even though I would have set the app to save to the SD card.
> I would need to download the one by one and each time wait before starting the next one -> that is in my opinion not great.

As soon as the first file finishes downloading, it is moved from the Internal Storage to the Shared Storage (the SD Card in this example). You should still be able to queue up all 4 books as long as you have enough storage for each individual file.

The issue would be that up to 5 files can be downloaded in parallel. I would need to double check whether this applies to all of the files in the queue or all of the files in the book (meaning the next book does not start until the previous book is finished).

jo-z commented:
just wanted to toss in a comment saying that downloads randomly stopped working for me on the main app, but work fine again on this one

nichwall committed:
Merge branch 'master' into android_download_rewrite

nichwall committed:
Inline `processDownloadItemParts`

nichwall committed:
Add internal storage size check for download item part

nichwall committed:
Add AbsLogger for download information

nichwall commented:
I added some additional logging and added a check to prevent trying to download files that are too large, but keep the files in the download queue so the files can still be downloaded in series instead of running out of space downloading in parallel. I have updated the top-level comment and marked 2 issues that I believe are fixed by this PR.

I believe this PR is ready for review and merging. Nobody has raised any new issues created by this PR in several GH linked issues or Discord and Reddit discussions.

The main issue people mention is background downloading does not work, but that did not work consistently before this PR anyway and this PR does not attempt to add that functionality.

advplyr committed:
Merge master

advplyr commented:
I started a download to internal storage then locked the screen half way through.
I know background downloading isn't supported but this caused an infinite loop.

Here
```kotlin
while (currentDownloadItemParts.isNotEmpty()) { }
```

<img width="1909" height="374" alt="image" src="https://github.com/user-attachments/assets/e922312f-d2cb-429a-b5ab-f14e81bfde7c" />

After closing and re-opening the app starting the download again resumes where it left off.

I'm able to reproduce that consistently when downloading to internal storage. 

When downloading to shared storage the download works in the background. I still have to go through all the code but from your description it sounds like they should function the same since shared storage is downloading to internal storage the same?

nichwall committed:
Change to coroutine scope

nichwall committed:
Download part removed after no updates for 10 seconds, still need to fix the UI

damccull commented:
Can we get another test build of this? And maybe see it merged in if it works? :D

damccull commented:
Ah, I figured out how to built it locally with 'act'.
For anyone else:
1) install docker + act
2) create a folder for artifacts
3) clone this repo
4) run:
    ```
    git fetch origin pull/1587/head:pr-1587
    git checkout pr-1587
    ```
5) run:
    ```
    act push --job build --artifact-server-path path/to/artifacts/folder
    ```
6) browse to artifacts folder
7) unzip the zip
8) untar the tar
9) copy the APK to your android and install it

damccull commented:
Would be cool to see a detection of internet return automatically and resume the download. Still gets stuck like the original, but when you kill and restart it, it continues from where it left off, so that's good.
